### PR TITLE
Expand attribute set for osu!standard SR calculation to assess length and combo

### PR DESCRIPTION
**Pull request by Xexxar**
# osu!Standard Star Rating Overhaul Proposal
###### Colloquially known as "joz sr"

Hello all.

On behalf of Joz (joseph-ireland), I am here the day of the new hotfix's deployment to propose a large change to the core of Star Rating (SR) calculation. This PR is _very_ large, and so this description will be a data analyst level overview, rather than an overly technical explanation. This change is the largest I've ever had to manage for something like this, so please bear with any errors on my part. With all that out of the way, let's start.

## The Problems with Existing Star Rating

While many might jump to say that the problem with the SR algorithm is the way skills assess difficulty, that is not the focus of this PR. This change is isolated to key fundamental problems in the methods of `Skill.cs` itself, and hopes to provide a solid foundation for future skill reworks that will address other large problems currently affecting osu!standard

### Combo

One of the biggest problems with PP that has affected standard since the beginning is the duality of combo and accuracy both playing a vital role in determining the skill required to perform a play. Under the current system, there is no way to properly gauge how impressive a non-FC play is, as we only have difficulty data available for the difficulty of the entire map. Currently, we scale combo by using the following algorithm:

```
            // Combo scaling
            if (Attributes.MaxCombo > 0)
                aimValue *= Math.Min(Math.Pow(scoreMaxCombo, 0.8) / Math.Pow(Attributes.MaxCombo, 0.8), 1.0);
```

It should be obvious that this is a band aid method of scaling due to the lack of any real data available for combo scaling. The problems with this are:

- If a player misses in the middle of the play, regardless of what they've FC'd before or after, their play is guttered.
- If a map has the hardest spike at the end or the beginning, a player can miss that part, but FC the rest and still be rewarded.
- This system encourages mappers and players to map and play shorter maps that players can maintain higher % combo on more easily

### Length and Difficulty Spikes

SR is calculated by first breaking a map into chunks, then arranging those chunks from hardest to easiest, and finally performing a geometric series with a base of `0.9`. While this method works, we come into a situation where the hardest 6 chunks, or 2.4 seconds, accounts for more than half of the star rating the map receives. Now, since the strain system builds upon itself, this doesn't necessarily mean that we're only judging 2.4 seconds of gameplay, but there are still issues with this concept. One is that the algorithm will generally under represent the SR of consistently difficult maps, while inflating the SR of spike-y maps where different sections have wildly differing difficulties 

The other issue with this system is the complete lack of any real length calculation inside of star rating. Because we assess difficulty with a geometric sum, after a certain point, additional chunks do not increase difficulty at all. To illustrate this, the hardest 29 chunks account for more than 95% of the difficulty of a map, regardless of the number of chunks that may exist beyond 29. It is for this reason that we have to account for length inside of PPcalc with the following formula:

```
            // Longer maps are worth more
            double lengthBonus = 0.95 + 0.4 * Math.Min(1.0, totalHits / 2000.0) +
                                 (totalHits > 2000 ? Math.Log10(totalHits / 2000.0) * 0.5 : 0.0);

            aimValue *= lengthBonus;
```

This again, showcases how clear the limitation of SR is in calculating length. Though objectively, every object is a chance to miss, we are effectively scaling PP by the number of objects while ignoring the difficulty of the object. This is very problematic, as you could have a map that is 90% streams with a few jumps, and the jumps are going to receive a huge buff to PP anyway due to the length of the map. To summarize, this systems problems are:

- Length being based on object count is arbitrary and doesnt account for the difficulty of those objects
- Geometric sum in SR calc encourages players and mappers to play and make maps with difficulty spikes
- The lack of a real way to assess the difficulty of a map beyond the hardest 20 seconds

## The Solutions

Introducing joz, the chadlad who made all of this possible. This again, is not going to be a technical dive into what the code does, rather it is an overview of the principles. Further technical documentation can be added as requested by developers. Future QOL changes are also going to be needed. This MR has been brought up to a moderately sophisticated level, and hopefully it has been documented well enough for most devs to understand.

### Strain Is the Inverse of the Probability of Missing

Currently, strain represents the difficulty of a section, but in a combo based system, every note is a chance of missing. It can be argued then that strain is effectively the skill needed to not miss, and thus, the inverse of strain can be viewed as the probability to not miss. What this new proposal does is extend this idea to the context of player skill. The probability of missing a certain note may be `.1` for player X's skill level, but is `.9` for player Y's skill. We can build a curve to represent this miss probability that has some nice propertied (detailed in the next section).

### Doubling Length Should Increase SR

The core idea of the new SR calc is based on the idea that doubling the length of a map while keeping the difficulty constant should increase the overall difficulty by a fixed percentage. I.E if you have a `6` star map and you copy paste the same map, it should be `6 * k`, where `k` is some constant (like `1.05`). A graph has been build below to demonstrate how this works in practice.

https://www.desmos.com/calculator/gk5ujznstf

In this graph, we have two groups of notes, one where every object has difficulty `d_1`, and one where every object has `d_2`. We can see that as we add notes to any group, the total SR goes up, and we can adjust the amount the SR goes up based on the `star_bonus_per_double` constant. This means that, as we add objects of varying difficulty, we can properly weight additional notes relative to the skill level of other notes.

### Storing SR Chunks as Attribute Data to Interpolate Non-FC SR

One of the perks in using the methods above is that we have a very handy property. If we take any two subsets of a map, we can add them together to get their combined difficulty. With this in mind, we can calculate the star rating for several sections and be able to derive the combined SR by adding any subset of these sections together. With that in mind, if someone FC'd 50% of a map, we could calculate an SR for 50% of the map by adding up those corresponding chunks.

Obviously though, a problem with this arises in that we do not know where the combo was achieved. In order to work around this, we assume the combo was achieved on the easiest possible section to get that combo on. If you FC'd the hardest 50% of the map, we will assume that you FC'd the easier 50% of the map. We do this by using an interpolation function. We can also use the same principle with miss data to ascertain whether or not more of the map might have been FC'd besides the single largest combo. 

However, in order to do this, we have to store these chunks as attribute data for every map. Currently this PR uses 20 sections, although I would propose we look at having this scale based on the maxCombo.

## Conclusions and Acknowledgements 

Because of how this system works, it is very important that we look at reworking skills. Due to the nature of a geometric decay rate based on MS for the process function, maps that have long jump patterns or heavy circle count will be overly inflated. This is seen in buffs on some undesirable maps. Likewise, we can see nerfs on maps that have high object count but are not consistently difficulty in the eyes of the strain system. Though this is partly intended, there is a balance that needs to be worked towards, which I believe is again where skill reworks come in.

We have created a website with test values with this build online ([link](https://xexxar.stanr.info)). This is by no means what we ultimately want the values to be, but is a demonstration of the system and can be used for further testing, performance or otherwise. With a change this large, it is likely much more important to view the change by itself to understand its purposes and refine it, then consider real diffCalc PP implementation using new skills and real balancing.

Thanks goes out to:

Joz (Joseph Ireland)
StanR
The ppDevDiscord

Thank you for reading.

### CODE REVIEW

**Comment by stanriders** [comment]
btw this is one of the biggest and most essential parts of https://github.com/ppy/osu/pull/9109

**Review by bdach — COMMENTED** [review]
This is the part of delta I haven't seen and reverse-engineered yet, so unfortunately I'll have to parse this in its entirety with no shortcuts. From a rough read-through I can already say code quality can be improved (`OsuSkill` in general can probably be split into smaller beings), and I can guess I can try doing that, but a few general remarks/questions before I proceed.

File: osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs [hunk 1]
```diff
@@ -26,6 +29,11 @@ public class OsuPerformanceCalculator : PerformanceCalculator
         private int countMeh;
         private int countMiss;
 
+        private const double miss_decay = 0.985;
```
[the remark this answers is not in the document]
Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/OsuPerformanceCalculator.cs:32]: Im guessing that this is a leftover from the old miss alg that was replaced by the new miss curve.

File: osu.Game.Rulesets.Catch/osu.Game.Rulesets.Catch.csproj [hunk 2]
```diff
@@ -5,6 +5,9 @@
     <AllowUnsafeBlocks>true</AllowUnsafeBlocks>
     <Description>catch the fruit. to the beat.</Description>
   </PropertyGroup>
+  <ItemGroup>
+    <PackageReference Include="MathNet.Numerics" Version="4.7.0" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (Xexxar) [reply osu.Game.Rulesets.Catch/osu.Game.Rulesets.Catch.csproj:10]: Joz did this. I wasnt sure what the convention is so I didnt change anything on my end. Ill remove anywhere its not needed.

File: osu.Game/Rulesets/Difficulty/Skills/Skill.cs [hunk 3]
```diff
@@ -16,7 +16,12 @@ namespace osu.Game.Rulesets.Difficulty.Skills
     public abstract class Skill
     {
         /// <summary>
-        /// The peak strain for each <see cref="DifficultyCalculator.SectionLength"/> section of the beatmap.
+        /// The length of each strain section.
+        /// </summary>
+        protected virtual int SectionLength => 400;
```
[the remark this answers is not in the document]
Comment (smoogipoo) [reply osu.Game/Rulesets/Difficulty/Skills/Skill.cs:21]: That PR is super complex to me for such a simple change. Can _just this_ be split out into its own PR? I think it's beneficial to have regardless of the rest of the stuff in #11407...

**Comment by bdach** [comment]
I've made like two-thirds of an initial understanding/clean-up pass of this. I've went through the processing bits, but not the calculations themselves yet. Seems mostly sane so far (and way *way* better readability wise than delta ever was). Still need to look at the part that spits out results.

From what I know the submitter is unavailable for the foreseeable future, but I'd promised I'd look at this and can try to spearhead this despite that.

File: osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs [hunk 4]
```diff
@@ -0,0 +1,411 @@
...
+            double fcSkill = fullComboDifficulty / fullComboDifficultyMultiplier;
+            double fcProb = fcProbability(fcSkill, fullComboDifficulty);
+
+            for (int i = 0; i < difficulty_count; ++i)
+            {
+                double missDifficulty = fullComboDifficulty * MISS_STAR_RATING_MULTIPLIERS[i];
+                double missSkill = missDifficulty / fullComboDifficultyMultiplier;
+
+                double[] missProbabilities = getMissProbabilities(missSkill);
+                result[i] = getMissCount(fcProb, missProbabilities);
```
Comment (bdach) [inline osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs:337]: This is one thing that seemed slightly weird in my pass - statistics was never my good side, but I'm not seeing the connection between the full combo probability and the per-note miss probabilities. The methodology (finding the zero of the CDF) looks correct, but I'm just not sure why this is the threshold chosen.

@joseph-ireland can you advise on this if you ever get a moment?

Comment (joseph-ireland) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs:337]: The full combo probability comes from the previous step - finding the skill that can FC with the given amount of time spent on failed attempts.

Using that as a threshold will make the 100% combo and the 0 miss values match up.

You could also interpret it as including the length of the map in there. 

successProbability = expectedTimePerPlay/expectedTimeUntilSuccess, so keeping it fixed is essentially assuming that the time spent per play of the map is the same for different skill levels, and they just allow themselves to make more misses.

Comment (bdach) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs:337]: Right, that does make sense, thanks!

Comment (Xexxar) [reply osu.Game.Rulesets.Osu/Difficulty/Skills/OsuSkill.cs:337]: I mentioned this to bdach in a PM, but I am still available for consulting in this thread. Taking a break primarily from the community relations aspect.

**Comment by Xexxar** [comment]
Closing this PR. 

Long term I still think we want to go with this type of implementation for SR and combo calculation: but for now we can use an approximation. Thanks to https://github.com/ppy/osu/pull/11690, with strain being opened up properly, we can write skills that approximate how joz's calculation works here to build better skills and balance them with performance calc. Effectively, we can get the length calculation benefits without the attribute changes. Once that is all done, we can look towards expanding the attribute set to give us better calculations of partial FCs, such as what this PR does. Skills rework will have more of an effect on the values than this PR, and is easier for the devs to implement atleast in the short term.

**Comment by stanriders** [comment]
I don't see any reason to close this especially if you want this to be a part of pp anyway. Attribute changes isn't something that's completely blocking development as far as I remember and using approximations while having a ready and well implemented algorithm is redundant unless you're looking at performance stats (which you never mention so I assume thats not the case). In my opinion it's better to finish up this PR instead of coming up with alternatives that provide less but take more time and effort in both development now and later down the line when we finally go back to joz's system.

**Pull request closed without merging**